gensim/models/nmf.py

```diff
@@ -0,0 +1,205 @@
+from itertools import chain
+
+import numpy as np
+from scipy.stats import halfnorm
+
+
+def _thresh(X, lambda1, vmax):
+    res = np.abs(X) - lambda1
+    np.maximum(res, 0.0, out=res)
+    res *= np.sign(X)
+    np.clip(res, -vmax, vmax, out=res)
+    return res
+
+
+def _mrdivide(B, A):
+    """Solve xB = A
+    """
+    if len(B.shape) == 2 and B.shape[0] == B.shape[1]:
+        return np.linalg.solve(B.T, A.T).T
+    else:
+        return np.linalg.lstsq(A.T, B.T, rcond=None)[0].T
+
+
+def _transform(W):
+    newW = W.copy()
+    np.maximum(newW, 0, out=newW)
+    sumsq = np.sqrt(np.sum(W**2, axis=0))
+    np.maximum(sumsq, 1, out=sumsq)
+    return _mrdivide(newW, np.diag(sumsq))
+
+
+def _solveproj(v, W, lambda1, kappa=1, h=None, r=None, v_max=None, max_iter=1e9):
+    m, n = W.shape
+    v = v.T
+    if v_max is None:
+        v_max = v.max()
+    if len(v.shape) == 2:
+        batch_size = v.shape[1]
+        rshape = (m, batch_size)
+        hshape = (n, batch_size)
+    else:
+        rshape = m,
+        hshape = n,
+    if h is None or h.shape != hshape:
+        h = np.zeros(hshape)
+
+    if r is None or r.shape != rshape:
+        r = np.zeros(rshape)
+
+    eta = kappa / np.linalg.norm(W, 'fro')**2
+
+    iters = 0
+
+    while True:
+        iters += 1
+        # Solve for h
+        htmp = h
+        h = h - eta * np.dot(W.T, np.dot(W, h) + r - v)
+        np.maximum(h, 0.0, out=h)
+
+        # Solve for r
+        rtmp = r
+        r = _thresh(v - np.dot(W, h), lambda1, v_max)
+
+        # Stop conditions
+        stoph = np.linalg.norm(h - htmp, 2)
+        stopr = np.linalg.norm(r - rtmp, 2)
+        stop = max(stoph, stopr) / m
+        if stop < 1e-5 or iters > max_iter:
+            break
+
+    return h, r
+
+
+class NMF:
```

Comment:
Use `SaveLoad` as baseclass (for save/load) + check `LdaModel` class (probably, parent class for lda will be useful here too).

Reply:
Done